example/src/main/java/io/netty/example/tunechoserver/InetUtils.java

```diff
@@ -0,0 +1,329 @@
+        sum = ((sum >> 16) & 0xFFFF) + (sum & 0xFFFF);
+        sum = ((sum >> 16) & 0xFFFF) + (sum & 0xFFFF);
+
+        // Negate the resultant sum to form the new checksum.
+        sum = ~sum & 0xFFFF;
+
+        return sum;
+    }
+
+    public static int updateInetChecksum(byte[] oldDataBuf, int oldDataOffset, byte[] newDataBuf,
+            int newDataOffset, int len, int oldSum) {
+
+        if (len > 65536) {
+            throw new IllegalArgumentException("checksum input too large");
+        }
+        if ((len & 1) != 0) {
+            throw new IllegalArgumentException("checksum input not multiple of two");
+        }
+
+        // NOTE: This algorithm is an adaptation of Equation 3 from RFC-1624.
+
+        // Start with the one's compliment negative of the previous checksum value.
+        int sum = ~oldSum & 0xFFFF;
+
+        // For each 2-byte word being updated...
+        for (; len > 1; oldDataOffset += 2, newDataOffset += 2, len -= 2) {
+
+            // Sum the one's compliment negative of the old word value.
+            sum += ~(((oldDataBuf[oldDataOffset] & 0xff) << 8) + (oldDataBuf[oldDataOffset + 1] & 0xff)) & 0xFFFF;
+
+            // Sum the new word value.
+            sum +=  ((newDataBuf[newDataOffset] & 0xff) << 8) + (newDataBuf[newDataOffset + 1] & 0xff);
+        }
+
+        // Complete the one's compliment sum by performing end-around carry.
+        sum = (sum & 0xFFFF) + ((sum >> 16) & 0xFFFF);
+        sum = (sum & 0xFFFF) + ((sum >> 16) & 0xFFFF);
+
+        // Negate the resultant sum to form the new checksum.
+        sum = ~sum & 0xFFFF;
+
+        return sum;
+    }
+
+    public static String ipVersionToString(int ipVersion) {
+        switch (ipVersion) {
+        case IP_VERSION_4:
+            return "IPv4";
+        case IP_VERSION_6:
+            return "IPv6";
+        default:
+            return Integer.toHexString(ipVersion).toUpperCase();
+        }
+    }
+
+    public static String ipProtocolToString(int ipProtocol) {
+        switch (ipProtocol) {
+        case IP_PROTOCOL_ICMP:
+            return "ICMP";
+        case IP_PROTOCOL_TCP:
+            return "TCP";
+        case IP_PROTOCOL_UDP:
+            return "UDP";
+        case IP_PROTOCOL_ICMPV6:
+            return "ICMPv6";
+        default:
+            return Integer.toHexString(ipProtocol).toUpperCase();
+        }
+    }
+
+    public static String ipAddressToString(byte[] addr) {
+        try {
+            return InetAddress.getByAddress(addr).getHostAddress();
+        } catch (Exception ex) {
+            throw new IllegalArgumentException("Not an IP address");
+        }
+    }
+
+    public static final class IPHeader {
+        public int ipVersion;
+        public int protocol;
+        public byte[] sourceAddress;
+        public byte[] destAddress;
+        public int headerLength;
+        public int payloadLength;
+        public int checksum;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make checksum a static final constant or non-public and provide accessors if needed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)